src/modules/power_profiles_daemon.cpp

```diff
@@ -0,0 +1,214 @@
+#else
+#include <fmt/core.h>
+#endif
+
+#include <glibmm.h>
+#include <glibmm/variant.h>
+#include <spdlog/spdlog.h>
+
+namespace waybar::modules {
+
+PowerProfilesDaemon::PowerProfilesDaemon(const std::string& id, const Json::Value& config)
+    : ALabel(config, "power-profiles-daemon", id, "{profile}", 0, false, true), connected_(false) {
+  if (config_["format"].isString()) {
+    format_ = config_["format"].asString();
+  } else {
+    format_ = "{icon}";
+  }
+
+  if (config_["tooltip-format"].isString()) {
+    tooltipFormat_ = config_["tooltip-format"].asString();
+  } else {
+    tooltipFormat_ = "Power profile: {profile}\nDriver: {driver}";
+  }
+  // Fasten your seatbelt, we're up for quite a ride. The rest of the
+  // init is performed asynchronously. There's 2 callbacks involved.
+  // Here's the overall idea:
+  // 1. Async connect to the system bus.
+  // 2. In the system bus connect callback, try to call
+  //    org.freedesktop.DBus.Properties.GetAll to see if
+  //    power-profiles-daemon is able to respond.
+  // 3. In the GetAll callback, connect the activeProfile monitoring
+  //    callback, consider the init to be successful. Meaning start
+  //    drawing the module.
+  //
+  // There's sadly no other way around that, we have to try to call a
+  // method on the proxy to see whether or not something's responding
+  // on the other side.
+
+  // NOTE: the DBus adresses are under migration. They should be
+  // changed to org.freedesktop.UPower.PowerProfiles at some point.
+  //
+  // See
+  // https://gitlab.freedesktop.org/upower/power-profiles-daemon/-/releases/0.20
+  //
+  // The old name is still announced for now. Let's rather use the old
+  // adresses for compatibility sake.
+  //
+  // Revisit this in 2026, systems should be updated by then.
+  Gio::DBus::Proxy::create_for_bus(Gio::DBus::BusType::BUS_TYPE_SYSTEM, "net.hadess.PowerProfiles",
+                                   "/net/hadess/PowerProfiles", "net.hadess.PowerProfiles",
+                                   sigc::mem_fun(*this, &PowerProfilesDaemon::busConnectedCb));
+}
+
+PowerProfilesDaemon::~PowerProfilesDaemon() {
+  if (powerProfileChangeSignal_.connected()) {
+    powerProfileChangeSignal_.disconnect();
+  }
+  if (powerProfilesProxy_) {
+    powerProfilesProxy_.reset();
+  }
+}
+
+void PowerProfilesDaemon::busConnectedCb(Glib::RefPtr<Gio::AsyncResult>& r) {
+  try {
+    powerProfilesProxy_ = Gio::DBus::Proxy::create_for_bus_finish(r);
+    using GetAllProfilesVar = Glib::Variant<std::tuple<Glib::ustring>>;
+    auto callArgs = GetAllProfilesVar::create(std::make_tuple("net.hadess.PowerProfiles"));
+
+    auto container = Glib::VariantBase::cast_dynamic<Glib::VariantContainerBase>(callArgs);
+    powerProfilesProxy_->call("org.freedesktop.DBus.Properties.GetAll",
+                              sigc::mem_fun(*this, &PowerProfilesDaemon::getAllPropsCb), container);
+    // Connect active profile callback
+  } catch (const std::exception& e) {
```

Comment:
`Glib::Error` does not inherit from `std::exception`, so you'll have to handle it in a separate `catch` branch for all DBus callbacks.

`util/format.hpp` has a formatter required for `Glib::Error::what()` result.